typeahead: Properly handle custom triggers.

aero31aero opened the pull request:
Fixes #14246.

This fixes an issue in our stream-topic typeahead code where sometimes typing `#den>` wouldn't complete the stream mention.

showell commented:
I wonder if this would actually be more maintainable if we just used `undefined` for `this.custom_selection_state` rather than having to give falsy values (of three different types!) for the three fields.

I would also make the commenting here slightly more verbose, as it's hard for somebody jumping in this code to know what's going on.

The variable name `custom_selection_state` is possibly confusing me here, because when I hear "selection" I think copy/paste instead of choosing, and I think this is more about choosing, correct?

showell commented:
I guess "selection" is already well explained in the big block comment (which I didn't see in the diff), so disregard that part of my prior comment.

aero31aero commented:
I was on the fence for using `selection` here, but the typeahead function is called `this.select()` so it made sense to follow the convention and say `trigger_selection`. Anyway, updated the default to be an empty object. This avoids having to write code like `if (this.x and this.x.y)` and also avoids having to specify any defaults.

> I would also make the commenting here slightly more verbose, as it's hard for somebody jumping in this code to know what's going on.

This is why I added a reference to the issue where I've written a larger write-up. Since the write-up was large enough, I figured it made sense to have it in a rendered markdown format instead of an in-file comment. (#14246).

To make sure people stumble on it, I added the comment at the top of the file as well now.

timabbott commented on static/js/composebox_typeahead.js:
```diff
@@ -752,7 +752,7 @@ exports.content_typeahead_selected = function (item, event) {
         }
         beginning += '#**' + item.name;
-        if (event && event.key === '>') {
+        if (this.custom_selection_state && this.custom_selection_state.key === '>') {
```
Can't we just pass a modified event that has the old format, or even just the `.key` value?  This interface feels weird.

aero31aero replied:
I could simplify this logic for readability. I'll push back to this PR with those changes.

timabbott commented on static/third/bootstrap-typeahead/typeahead.js:
```diff
@@ -352,7 +368,8 @@
         default:
-          if (this.trigger_selection(e)) {
+          if (this.custom_selection_state.should_trigger) {
+            this.custom_selection_state.should_trigger = false;
```
This code is kinda hard to follow.  Are we confident that `this.custom_selection_state` is always set during a `keyup` call?  And is the `keyCode` in `custom_selection_state` used for anything?

aero31aero replied:
Yes, a keyup event implies that we have already had a keypress event.
